README.md
# Contao-OM-Backend

## Introduction

Contao-OM-Backend is a module for the Open Source PHP Content Management System Contao. It was designed for admins and daily work.

## System requirements

- Contao 4.4.x+

## Installation

You can install the module through composer.

## Features (in german)

- __Toolbar:__ praktische responsive Toolbar mit Buttons für wiederkehrende Aufgaben (Benutzer-Einstellungen) 
- __Sprache bei Startpunkten:__ zeigt zusätzlich die ausgewählte Sprache bei den Startpunkten an (Benutzer-Einstellungen)
- __ID-/Alias-Suche:__ ein eigenständiges Backend-Modul was nach IDs und Aliases in allen tl_*-Tabellen suchen kann
- __IDs bei Shift:__ zeigt bei gedrückter Shift-Taste in einigen Listenansichten zusätzlich die IDs der Datensätze an (Benutzer-Einstellungen)
- __Textfeld-Länge:__ zeigt bei allen DCA-Feldern vom Typ 'text', die derzeitige Länge und falls vorhanden auch die maximale Länge an
- __MarkdownViewer:__ Backend-Modul zur Anzeige aller Markdown-Dateien im Wurzelverzeichnis der Installation als HTML (Benutzer-Einstellungen)
- __Backend-CSS:__ bringt neue Klassen für die Evaluationseigenschaft 'tl_class' mit, etwa w25,w33,w66,w75,heightAuto
- __Backend-Tabs:__ Reiter-Ansicht für eigenen Backend-Module
- __Backend-Links:__ Backend-Modul zur Erstellung für Kurzlinks im Backend (Haupt-Navigation und Top-Navigation)
- __User-Redirect:__ Backendbenutzer bei Login innerhalb des Backends weiterleiten
- __Layout-Buttons:__ Layout-Button in Seitenstruktur (inkl. Vererbung)
- __Speichern-Buttons:__ die Untermenüs werden wieder aufgelöst
- __Full-Width_Backend:__ Backend wieder auf volle Breite bringen
- __Element-Klassen:__ Einzelnen Inhaltselementen oder Artikeln CSS-Klassen zuweisen
- __Reihenfolge der Backend-Module:__ Einstellbar in den Systemeinstellungen
- __Neueste Contao-Version anzeigen:__ Systemeinstellungen

For more informations read FEATURES.md.

## Planned features (in german)

- __Suche nach Datei-Verwendungen:__ kann bei der Suche nach Verwendung von Dateien helfen (Dateiverwaltung » Datei bearbeiten; Hinweise in FEATURES.md beachten)

## Missing informations

You missed information? Send me an mail rene.fehrmann@omos.de!

## License

Contao-OM-Backend is licensed under the terms of the LGPLv3. The full license text is
available at [`gnu.org`][1].

Note that the LGPL incorporates the terms and conditions of the GPL, therefore
both licenses are included there. This, however, does not imply that OM-Backend is
dual licensed under both the GPL and the LGPL.

[1]: http://www.gnu.org/licenses/lgpl-3.0.de.html

CHANGELOG.md
# Contao bundle change log

### 1.7.5 (2023-04-17)

 + Fixed: Wrong backend user detection

### 1.7.4 (2023-03-28)

 + Fixed: Route prefix detection in Contao 4.9 defect
 + Fixed: One more php 8 issue

### 1.7.3 (2023-03-21)

 + Fixed: Some PHP 8 errors (thx @fritzmg)
 + Fixed: Add some missing parameter handling for `contao.backend.route_prefix`
 + Fixed: Use parameter `contao_manager.manager_path` and `contao.web_dir` for manager detection

### 1.7.2 (2023-02-03)

 + Fixed: Some PHP 8 errors
 + Changed: Deactivate feature `Backend-Tabs` (please remove your this callback in your config.php)

### 1.7.1 (2021-12-17)

 + Fixed: Some PHP warnings

### 1.7.0 (2021-12-16)

 + Improvements for PHP 8
 + Increase system requirements

### 1.6.14 (2021-11-21)

 + Added: Version 4.13 as known lts
 + Added: Replace insert tags in markdown viewer 

### 1.6.13 (2021-05-02)

 + Fixed: Missing if condition on version check

### 1.6.12 (2021-04-28)

 + Fixed: Remove dependency for "wa72/htmlpagedom"

### 1.6.11 (2021-02-12)

 + Fixed: User redirect not working

### 1.6.10 (2021-02-09)

 + Fixed: Backendlinks in Contao 4.9+ not working
 + Fixed: Problem with user rights in Contao 4.9+
 + Fixed: Redirect token problem with id search

### 1.6.9 (2020-09-02)

 + Fixed: Wrong version compare

### 1.6.8 (2020-08-25)

 + Fixed: Overwrite CSS for all links only on backend module sysinfo phpinfo
 + Fixed: Wrong alignment for language on root pages
 
### 1.6.7 (2020-05-22)

 + Added: Add name of layout to title on layout button (see #44)
 + Changed: Replace 'menatwork/contao-multicolumnwizard' with 'menatwork/contao-multicolumnwizard-bundle' (see #34)
 + Fixed: Version feature breaks version info in sysinfo tab packages (see #46)
 + Fixed: Version info shows lower version than current (see #47) 
 + Fixed: Missing error message if no md-File found (see #43)
 + Fixed: Can´t use install tool because redirect to login

### 1.6.6 (2019-12-11)

 + Added: Support rsce-Templates for feature 'Element classes' (see #41)

### 1.6.5 (2019-12-07)

 + Fixed: Hook 'parseTemplate' for element classes not called in frontend (#41) 

### 1.6.4 (2019-11-09)

 + Fixed: Use of 'return;' in config/config.php breaks other extension includes

### 1.6.3 (2019-10-29)

 + Fixed: Do not add features when on install tool
 + Fixed: Some issues with the monolog.logger

### 1.6.2 (2019-10-27)

 + Changed: Show latest versions only if different to current version
 + Added: Log entry for getting contao versions from github

### 1.6.1 (2019-10-18)

 + Fixed: Some issues with version feature

### 1.6.0 (2019-10-18)

 + Added: Show newest Contao version in backend (System settings)

### 1.5.2 (2019-08-31)

 + Fixed: Consider the rights management on create backend module order

### 1.5.1 (2019-07-15)

 + Fixed: Solve an array problem with feature 'Order of backend modules'
 + Fixed: Add missing feature to README.md

### 1.5.0 (2019-07-14)

 + Added: New option to set the order of the backend modules in the system settings

### 1.4.2 (2019-05-22)

 + Changed: Function id-view now searchable and colored, click again shift to remove
 + Changed: Remove dca field 'cssClasses' if no classes assigned
 + Fixed: Replace some used deprecated functions

### 1.4.1 (2019-03-17)

 * Fixed: Backend link problems
 * Fixed: Tables are not correct added to backend modules when use backend tabs
 * Fixed: Some backend link problems with app_dev.php

### 1.4.0 (2019-03-16)

 * Added: Make markdown file editable
 * Added: New feature sysinfo
 * Fixed: Correct language array for id-search

### 1.3.3 (2018-07-14)

 * Fixed: Error feature module-tabs is used without own tables in backend module

### 1.3.2 (2018-07-10)

 * Fixed: A foreach problem with module-tabs 
 * Changed: Remove the feature fullwidth from 4.5, since it is no longer necessary

### 1.3.1 (2018-07-02)

 * Fixed: Theme buttons with wrong id´s (then multiple themes exists)
 * Fixed: Illegal string warning on rocksolid custom elements
 * Fixed: Search contao-manager.phar.php in wrong directory
 * Fixed: Dom manipulation has a problem with the markup from the mediamanager from isotope.
 * Fixed: Problem with script tag in allowed HTML tags / settings
 * Fixed: HTML-Entitäten werden in den Übersetzungen wieder dekodiert
 * Fixed: Add missing language labels

### 1.3.0 (2018-01-13)

 * Added: New feature 'Element classes'
 * Fixed: 'Call to a member function hasAccess() on null' on site structure page
 * Fixed: Add 'rel="noopener"' to links with 'target="_blank"'

### 1.2.3 (2017-12-11)

 * Added: Option '_blank' for backend links 
 * Fixed: A problem with BackendUser->authenticate() and the install tool on fresh installations

### 1.2.2 (2017-12-02)

 * Fixed: Wrong paddings of top links
 * Fixed: Missing tooltips for edit multiple buttons in the toolbar
 * Fixed: Edit button not visible, with option 'save buttons'

### 1.2.1 (2017-11-28)

 * Fixed: Array errors (Thanks to Tastaturberuf)
 * Fixed: Some stuff find by phpstorm inspections

src/Resources/contao/classes/Toolbar.php
<?php

/**
 * Contao bundle contao-om-backend
 *
 * @copyright OMOS.de 2017 <http://www.omos.de>
 * @author    René Fehrmann <rene.fehrmann@omos.de>
 * @link      https://github.com/OMOSde/contao-om-backend
 * @license   LGPL 3.0+
 */


/**
 * Namespace
 */
namespace OMOSde\ContaoOmBackendBundle;


/**
 * Usages
 */

use Contao\System;
use Contao\ThemeModel;


/**
 * Class Toolbar
 *
 * @copyright OMOS.de 2017 <http://www.omos.de>
 * @author    René Fehrmann <rene.fehrmann@omos.de>
 */
class Toolbar extends \Backend
{
    private string $routePrefix;

    /**
     * Toolbar constructor.
     */
    public function __construct()
    {
        parent::__construct();
        $this->import('BackendUser', 'User');

        $container = System::getContainer();
        $this->routePrefix = $container->hasParameter('contao.backend.route_prefix') ? $container->getParameter('contao.backend.route_prefix') : '/contao';
    }


    /**
     * Adds a toolbar to the backend template
     *
     * @param $strContent
     * @param $strTemplate
     *
     * @return mixed
     */
    public function addToolbarToBackendTemplate($strContent, $strTemplate)
    {
        if ($strTemplate == 'be_main' && is_array($this->User->om_backend_features) && in_array('addToolbar', $this->User->om_backend_features))
        {
            // add new css class to body
            $strContent = str_replace('<body id="top" class="', '<body id="top" class="om_backend_toolbar ', $strContent);

            // add html
            $strContent = str_replace('<div id="container"', $this->generateToolbar($strContent).'<div id="container"' , $strContent);
        }

        return $strContent;
    }


    /**
     * Generate the html for the toolbar
     *
     * @param $strContent
     * @return string
     */
    protected function generateToolbar($strContent)
    {
        // handle entry point app_dev.php
        $strEntryPoint = (strpos(\Environment::get('request'), 'app_dev.php') !== false) ? 'app_dev.php/' : '';

        // get a request token from csrf service
        $container = \System::getContainer();
        $strToken = $container->get('security.csrf.token_manager')->getToken($container->getParameter('contao.csrf_token_name'))->getValue();

        // start html
        $strToolbar  = '<div id="om_backend_toolbar">';

        // add buttons id-search, install-tool, new template, sync files
        $strToolbar .= sprintf('<a class="button" href="%s?do=id_search" title="%s">%s</a>',
            $strEntryPoint.$this->routePrefix,
            $GLOBALS['TL_LANG']['om_backend']['id_search'],
            $this->getIcon('idSearch')
        );
        $strToolbar .= sprintf('<a class="button" href="%s/install" target="_blank" rel="noopener" title="%s">%s</a>',
            $strEntryPoint.$this->routePrefix,
            $GLOBALS['TL_LANG']['om_backend']['install_tool'],
            $this->getIcon('installTool')
        );
        $strToolbar .= sprintf('<a class="button" href="%s?do=tpl_editor&key=new_tpl&rt=%s" title="%s">%s</a>',
            $strEntryPoint.$this->routePrefix,
            $strToken,
            $GLOBALS['TL_LANG']['om_backend']['new_template'],
            $this->getIcon('newTemplate')
        );
        $strToolbar .= sprintf('<a class="button" href="%s?do=files&act=sync&rt=%s" title="%s">%s</a>',
            $strEntryPoint.$this->routePrefix,
            $strToken,
            $GLOBALS['TL_LANG']['om_backend']['sync_files'],
            $this->getIcon('syncFiles')
        );

        // add contao manager icon
        if ($this->contaoManagerExists())
        {
            $strToolbar .= sprintf('<a class="button" href="%s" title="%s" target="_blank" rel="noopener">%s</a>',
                System::getContainer()->getParameter('contao_manager.manager_path'),
                $GLOBALS['TL_LANG']['om_backend']['contao_manager'],
                $this->getIcon('contaoManager')
            );
        }

        // add sections stylesheets, modules, layouts, image_sizes from themes
        $objThemes = ThemeModel::findAll(array('order'=>'name'));
        if ($objThemes)
        {
            foreach ($objThemes as $theme)
            {
                // add separator
                $strToolbar .= '<div class="separator"></div>';

                // add buttons
                $strToolbar .= sprintf('<a class="button" href="%s?do=themes&amp;table=tl_style_sheet&amp;id=%s&amp;rt=%s" title="%s">%s</a>',
                    $strEntryPoint.$this->routePrefix,
                    $theme->id,
                    $strToken,
                    sprintf('%s (%s)', $GLOBALS['TL_LANG']['om_backend']['stylesheets'], $theme->name),
                    $this->getIcon('themeStylesheet')
                );
                $strToolbar .= sprintf('<a class="button" href="%s?do=themes&amp;table=tl_module&amp;id=%s&amp;rt=%s" title="%s">%s</a>',
                    $strEntryPoint.$this->routePrefix,
                    $theme->id,
                    $strToken,
                    sprintf('%s (%s)', $GLOBALS['TL_LANG']['om_backend']['modules'], $theme->name),
                    $this->getIcon('themeModule')
                );
                $strToolbar .= sprintf('<a class="button" href="%s?do=themes&amp;table=tl_layout&amp;id=%s&amp;rt=%s" title="%s">%s</a>',
                    $strEntryPoint.$this->routePrefix,
                    $theme->id,
                    $strToken,
                    sprintf('%s (%s)', $GLOBALS['TL_LANG']['om_backend']['layouts'], $theme->name),
                    $this->getIcon('themeLayout')
                );
                $strToolbar .= sprintf('<a class="button" href="%s?do=themes&amp;table=tl_image_size&amp;id=%s&amp;rt=%s" title="%s">%s</a>',
                    $strEntryPoint.$this->routePrefix,
                    $theme->id,
                    $strToken,
                    sprintf('%s (%s)', $GLOBALS['TL_LANG']['om_backend']['image_size'], $theme->name),
                    $this->getIcon('themeImageSize')
                );
            }
        }

        // generate save buttons
        if (strpos($strContent, 'class="tl_submit_container"') !== false && strpos($strContent, 'name="save"') !== FALSE)
        {
            $strSave = '';
            $arrButtons = ['save', 'saveNclose', 'saveNcreate', 'saveNduplicate', 'saveNback'];

            foreach ($arrButtons as $button)
            {
                if (strpos($strContent, 'name="'.$button.'"') !== FALSE)
                {
                    $strSave .= sprintf('<a class="button" onclick="document.getElementById(\'%s\').click(); return false;" title="%s">%s</a>',
                        $button,
                        $GLOBALS['TL_LANG']['MSC'][$button],
                        $this->getIcon($button)
                    );
                }
            }
            $strToolbar .= (strlen($strSave)) ?  '<div class="separator"></div>'.$strSave : '';
        }

        // add edit multiple
        if (strpos($strContent, 'class="header_edit_all"') !== false)
        {
            // get parameter
            $id    = (strlen(\Input::get('id'))) ? '&amp;id'.\Input::get('id') : '';
            $table = (strlen(\Input::get('table'))) ? '&amp;table='.\Input::get('table') : '';

            // add separator and button
            $strToolbar .= '<div class="separator"></div>';
            $strToolbar .= sprintf('<a class="button" href="%s?do=%s&amp;act=select&amp;rt=%s" title="%s">%s</a>',
                $strEntryPoint.$this->routePrefix,
                \Input::get('do').$table.$id,
                $strToken,
                $GLOBALS['TL_LANG']['om_backend']['stylesheets'].$objThemes->name,
                $this->getIcon('editMultiple')
            );
        }

        // edit multiple buttons
        if (strpos($strContent, 'class="tl_submit_container"') !== false && strpos($strContent, 'name="edit"') !== false)
        {
            // variables
            $arrButtons = [];

            // html button names
            $arrButtonNames = array('delete', 'cut', 'copy', 'override', 'edit', 'alias');

            // check for buttons
            foreach ($arrButtonNames as $buttonName)
            {
                // button save
                if (strpos($strContent, 'name="'.$buttonName.'"') !== false)
                {
                    $arrButtons[] = $buttonName;
                }
            }

            // add alls buttons and separator
            if (count($arrButtons) > 0)
            {
                // add separator
                $strToolbar .= '<div class="separator"></div>';

                // add buttons
                foreach ($arrButtons as $button)
                {
                    $strToolbar .= sprintf('<a class="button" onclick="document.getElementById(\'%s\').click(); return false;" title="%s">%s</a>',
                        $button,
                        $GLOBALS['TL_LANG']['om_backend']['button_'.$button],
                        $this->getIcon('multiple'.$button)
                    );
                }
            }
        }

        // create template button
        if (\Input::get('do') == 'tpl_editor' && \Input::get('key') == 'new_tpl')
        {
            // add separator
            $strToolbar .= '<div class="separator"></div>';

            // add button
            $strToolbar .= sprintf('<a class="button" onclick="document.getElementById(\'create\').click(); return false;" title="%s">%s</a>',
                $GLOBALS['TL_LANG']['om_backend']['button_save'],
                $this->getIcon('save')
            );
        }

        // close container
        $strToolbar .= '</div>';

        return $strToolbar;
    }


    /**
     * Check contao-manager.phar.php in document root
     *
     * @return bool
     */
    protected function contaoManagerExists()
    {
        $managerPath = System::getContainer()->getParameter('contao_manager.manager_path') ?? 'contao-manager.phar.php';
        $webDir = System::getContainer()->getParameter('contao.web_dir');

        return file_exists($webDir.'/'.$managerPath);
    }


    /**
     * @param $strIconName
     *
     * @return string
     */
    protected function getIcon($strIconName)
    {
        switch ($strIconName)
        {
            case 'idSearch':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><circle cx="10.5" cy="10.5" r="7.5"/><line x1="21" y1="21" x2="15.8" y2="15.8"/></svg>';

            case 'installTool':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><rect x="2" y="2" width="20" height="8" rx="2" ry="2" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><rect x="2" y="14" width="20" height="8" rx="2" ry="2" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="6" y1="6" x2="6" y2="6" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="6" y1="18" x2="6" y2="18" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'newTemplate':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><polyline points="14 2 14 8 20 8" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="12" y1="18" x2="12" y2="12" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="9" y1="15" x2="15" y2="15" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'syncFiles':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><polyline points="1 4 1 10 7 10"/><polyline points="23 20 23 14 17 14"/><path d="M20.49 9A9 9 0 0 0 5.64 5.64L1 10m22 4l-4.64 4.36A9 9 0 0 1 3.51 15"/></svg>';

            case 'themeLayout':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><rect x="3" y="3" width="18" height="18" rx="2" ry="2"/><line x1="3" y1="9" x2="21" y2="9"/><line x1="9" y1="21" x2="9" y2="9"/></svg>';

            case 'themeImageSize':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><rect x="3" y="3" width="18" height="18" rx="2" ry="2"/><circle cx="8.5" cy="8.5" r="1.5"/><polyline points="21 15 16 10 5 21"/></svg>';

            case 'themeModule':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><circle cx="12" cy="12" r="3" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><path d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z" fill="none" stroke="#91979c" stroke-miterlimit="10" stroke-width="2"/></svg>';

            case 'themeStylesheet':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><circle cx="12" cy="12" r="10" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><polygon points="16.24 7.76 14.12 14.12 7.76 16.24 9.88 9.88 16.24 7.76" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'editMultiple':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M20 14.66V20a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h5.34"/><polygon points="18 2 22 6 12 16 8 16 8 12 18 2"/></svg>';

            case 'save':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><polyline points="17 21 17 13 7 13 7 21" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><polyline points="7 3 7 8 15 8" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'saveNcreate':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="12" y1="9" x2="12" y2="15" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="9" y1="12" x2="15" y2="12" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'saveNclose':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="9" y1="9" x2="15" y2="15" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="9" y1="15" x2="15" y2="9" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'saveNduplicate':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><rect x="7" y="7" width="6" height="6" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><rect x="11" y="11" width="6" height="6" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'saveNback':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><line x1="9" y1="12" x2="15" y2="12" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/><polyline points="12 15 9 12 12 9" fill="none" stroke="#91979c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"/></svg>';

            case 'multipledelete':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M21 4H8l-7 8 7 8h13a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2z"></path><line x1="18" y1="9" x2="12" y2="15"></line><line x1="12" y1="9" x2="18" y2="15"></line></svg>';

            case 'multiplecut':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><circle cx="6" cy="6" r="3"></circle><circle cx="6" cy="18" r="3"></circle><line x1="20" y1="4" x2="8.12" y2="15.88"></line><line x1="14.47" y1="14.48" x2="20" y2="20"></line><line x1="8.12" y1="8.12" x2="12" y2="12"></line></svg>';

            case 'multiplecopy':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><rect x="9" y="9" width="13" height="13" rx="2" ry="2"></rect><path d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"></path></svg>';

            case 'multipleoverride':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><polygon points="14 2 18 6 7 17 3 17 3 13 14 2"></polygon><line x1="3" y1="22" x2="21" y2="22"></line></svg>';

            case 'multipleedit':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><polygon points="16 3 21 8 8 21 3 21 3 16 16 3"></polygon></svg>';

            case 'multiplealias':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><line x1="12" y1="6" x2="12" y2="18"/><line x1="3" y1="9" x2="8" y2="9"/><line x1="3" y1="13" x2="8" y2="13"/><line x1="16" y1="9" x2="21" y2="9"/><line x1="16" y1="13" x2="21" y2="13"/></svg>';

            case 'contaoManager':
                return '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="#91979c" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><path d="M3 9l9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"></path><polyline points="9 22 9 12 15 12 15 22"></polyline></svg>';
        }

        return '';
    }
}
